semanticcpg/src/main/scala/io/shiftleft/semanticcpg/utils/ExternalCommand.scala

```diff
@@ -1,3 +1,130 @@
+      case 0 => Some(stdOut)
+      case _ => None
+    }
+
+    def toTry: Try[Seq[String]] = exitCode match {
+      case 0 => Success(stdOut)
+      case nonZeroExitCode =>
+        val allOutput = stdOut ++ stdErr
+        val message = s"""Process exited with code $nonZeroExitCode. 
+                         |Input: $commandInfo
+                         |Output:
+                         |${allOutput.mkString(System.lineSeparator())}
+                         |""".stripMargin
+        Failure(new RuntimeException(message))
+    }
+  }
+
+  def run(
+    command: Seq[String],
+    workingDir: Option[String] = None,
+    mergeStdErrInStdOut: Boolean = false,
+    extraEnv: Map[String, String] = Map.empty,
+    isShellCommand: Boolean = false,
+    timeout: Duration = Duration.Inf
+  ): ExternalCommandResult = {
+    val cmd =
+      if (isShellCommand) {
+        val invokeShell =
+          if (isWin) Seq("cmd.exe", "/C")
+          else Seq("sh", "-c")
+        invokeShell ++ command
+      } else {
+        command
+      }
+
+    val builder = new ProcessBuilder().command(cmd.toArray*)
+    builder.environment().putAll(extraEnv.asJava)
+    builder.redirectErrorStream(mergeStdErrInStdOut)
+    workingDir.map(new File(_)).foreach(builder.directory)
+
+    val stdOutFile = File.createTempFile("x2cpg", "stdout")
+    val stdErrFile = Option.when(!mergeStdErrInStdOut)(File.createTempFile("x2cpg", "stderr"))
+    builder.redirectOutput(stdOutFile)
+    stdErrFile.foreach(builder.redirectError)
+
+    val commandInfo = s"""cmd: `${cmd.mkString(" ")}`, workingDir: ${workingDir.getOrElse(
+        System.getProperty("user.dir")
+      )}, extraEnv: $extraEnv"""
+    try {
+      logger.debug(s"executing command: $commandInfo")
+      val process = builder.start()
+      if (timeout.isFinite) {
+        logger.debug(s"waiting until command completes (but max $timeout)")
+        val finished = process.waitFor(timeout.toMillis, TimeUnit.MILLISECONDS)
+        if (!finished) {
+          logger.warn(s"timeout reached - will now kill the external command")
+          process.destroy()
+          throw new TimeoutException(s"command '${command.mkString(" ")}' with timeout='$timeout' has timed out")
+        }
+      } else {
+        logger.debug("waiting (indefinitely) until command completes")
+        process.waitFor()
+      }
+
+      logger.debug(s"command finished successfully")
+      val stdOut = IOUtils.readLinesInFile(stdOutFile.toPath)
+      val stdErr = stdErrFile.map(_.toPath).map(IOUtils.readLinesInFile).getOrElse(Seq.empty)
+      ExternalCommandResult(process.exitValue(), stdOut, stdErr, commandInfo)
+    } catch {
+      case NonFatal(exception) =>
+        val stdOut = IOUtils.readLinesInFile(stdOutFile.toPath)
+        val stdErr = stdErrFile.map(f => IOUtils.readLinesInFile(f.toPath)).getOrElse(Seq.empty) :+ exception.getMessage
+        logger.warn(s"command did not finish successfully. Input was: $commandInfo")
```

Comment:
Just a small remark here:
We frequently use `ExternalCommand.run` to check if a (compatible) executable FOOastgen binary is in the path by just trying to execute it (various frontends do so). Iff not, we call the one from the Joern distribution.
This results in a WARN log entry every time. 
I am not sure whats the best option here but maybe we could use an additional parameter that disables that warning in case you kind of expect the execution to fail (most people do not have the astgen binary in their PATH but some do)?

Reply:
Good point. I guess when I added that code (prior to this PR) I thought it's ok because users can configure their log levels accordingly. Rethinking, I guess it's better to not log here at all, and leave the handling of the result to the caller. Will change.